[TypeChecker] Implement limited set of conversions between Swift and C pointers

xedin opened the pull request:
Allow following pointer conversions in argument positions to C/ObjC imported calls:

- Allow implicit conversion to `Unsafe[Mutable]Pointer<UInt8>` or
   `Unsafe[Mutable]Pointer<Int8>` when the argument type is either
   `Unsafe[Mutable]Pointer<T>` or `Unsafe[Mutable]RawPointer`.

- Allow implicit conversion between signed and unsigned variants of trivial integer types:
  e.g. `Unsafe[Mutable]Pointer<Int64>` to/from `Unsafe[Mutable]Pointer<UInt64>`

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves: SR-10246

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Add a new conversion - Swift to C pointers

Following pointer conversions are supported in argument positions (when referencing C/ObjC functions):

- Unsafe[Mutable]RawPointer -> Unsafe[Mutable]Pointer<[U]Int>
- Unsafe[Mutable]Pointer<Int{8, 16, ...}> -> Unsafe[Mutable]Pointer<UInt{8, 16, ...}>

xedin committed:
[ConstraintSystem] Add a helper to determine whether locator is for argument to C/ObjC imported decl

xedin committed:
[ConstraintSystem] Implement new Swift -> C pointer conversions

Allow following conversions in argument positions
(applies only to call to imported C/ObjC declarations):

- Unsafe[Mutable]RawPointer -> Unsafe[Mutable]Pointer<[U]Int>
- Unsafe[Mutable]Pointer<Int{8, 16, ...}> <-> Unsafe[Mutable]Pointer<UInt{8, 16, ...}>

xedin committed:
[ConstraintSystem] NFC: Add test-cases for interactions between Swift and C pointers

xedin commented:
@swift-ci please build toolchain

xedin committed:
[CSSimplify] Allow implicit Swift -> C pointer conversion within optional types

Support Swift -> C pointer conversions even if argument required
a value to optional promotion or is optional.

xedin committed:
[TypeChecker] NFC: Add test-cases for Swift -> C pointer conversions

Make sure that conversions are properly supported when arguments
either is wrapped in an optional type or requires a value-to-optional
promotion because parameter is optional pointer.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test macOS platform

atrick commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5785,6 +5785,43 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
                                    ConversionRestrictionKind::PointerToPointer);
               }
             }
+
```
It looks to me like the ~150 lines inside the `getAnyPointerElementType` check would be nicely separable in a function (as a separate cleanup)

atrick replied:
Or, at least the part that's specific to `ConstraintKind::ArgumentConversion`

xedin replied:
Sure, I think it would be a good exercise to try to factor all the logic into restriction kinds e.g. `simplifyRestriction(...)` just like we do with constraints. Something I wanted to do but never had time...

atrick commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5785,6 +5785,43 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+                    ConversionRestrictionKind::PointerToCPointer);
+              }
+
+              // Unsafe[Mutable]Ponter -> Unsafe[Mutable]Pointer
```
Typo: s/Ponter/Pointer

xedin replied:
Thanks, I'm going to fix this when I rebase.

xedin replied:
Fixed!

atrick commented on test/Constraints/swift_to_c_pointer_conversions.swift.gyb:
```diff
@@ -0,0 +1,116 @@
+  % end
+}
+
+%for TestPtrSize in ['16', '32', '64']:
```
I think we should also have tests for passing `Unsafe[Mutable]Pointer<[U]Int8>` as an argument:
to `char *` - ok
to `int16-64` - error

xedin replied:
Ugh, for some reason I have completely ignored 8-bit integers...

xedin replied:
Added `8` bit size to both tests blocks, so everything should be covered now.

atrick replied:
ok, looks good. I think we should test that `UP<[U]Int8>` cannot be converted to other integer sizes, e.g. `UP<[U]Int32>`. Just to be clear that it is intentionally not supported. Someone might expect that we meant to support that case, but it's only safe to do that conversion in one direction. We're only giving [U]Int8 special aliasing powers on the C side of the call.

xedin replied:
That's what the tests down the line do e.g. `uint_${TestPtrSize}_ptr_func(${pointer})`, gyb doesn't make it easy to read but `Int8` is tested against all the signed and unsigned integer types of 16, 32, 64 and all `*_char_ptr_func` test it against Int8 and UInt8 pointers at the top of the test function.

atrick replied:
Oh right... I was just looking at the diff. So I can't think of anything else to test then.

xedin replied:
No worries!

xedin committed:
[TypeChecker] NFC: Test Swift -> C pointer conversions for [U]Int8 pointers

xedin committed:
[TypeChecker] NFC: Enable Swift -> C implicit pointer conversions tests only on Darwin

xedin commented:
@swift-ci please clean test

xedin committed:
[ConstraintSystem] NFC: Extract Swift -> C pointer conversion simplification into a separate method

xedin committed:
[ConstraintSystem] De-prioritize solutions with Swift -> C pointer conversions

Just like other implicit conversions - always prefer solutions with
the lowest possible number of them.

xedin commented:
@swift-ci please build toolchain

xedin committed:
[CSFix] Add a tailored fix for invalid Swift -> C pointer conversions

This is a skeleton of a fix that would be used to diagnose situations
when Swift -> C pointer conversion was attempted on a Swift function.

xedin committed:
[Diagnostics] Add a tailored diagnostic for Swift -> C pointer conversion

Diagnose situations where Swift -> C pointer implicit conversion
is attempted on a Swift function instead of one imported from C header.

```swift
func test(_: UnsafePointer<UInt8>) {}

func pass_ptr(ptr: UnsafeRawPointer) {
  test(ptr) // Only okay if `test` was an imported C function.
}
```

xedin committed:
[CSSimplify] Make sure that Swift -> C pointer conversion always waits for optionals

Wait for a value-to-optional promotion or optional-to-optional conversion
to happen before attempting Swift -> C pointer conversion.

xedin committed:
[CSSimplify] Detect and diagnose Swift -> C pointer conversion failures

Detect and diagnose situations when Swift -> C pointer conversion
is unsupported due to method not being imported from C header.

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please smoke test
